src/shared/GitLab/GitLabHostProvider.cs

```diff
@@ -0,0 +1,260 @@
+                }
+                else
+                {
+                    Context.Trace.WriteLine($"Invalid value for supported authentication modes override setting: '{authModesStr}'");
+                }
+            }
+
+            if (GitLabConstants.IsGitLabDotCom(targetUri))
+            {
+                return AuthenticationModes.All;
+            }
+
+            Context.Streams.Error.WriteLine($"Missing OAuth configuration for {targetUri.Host}, see https://github.com/GitCredentialManager/git-credential-manager/blob/main/docs/gitlab.md.");
+            // Would like to query password_authentication_enabled_for_git, but can't unless logged in https://gitlab.com/gitlab-org/gitlab/-/issues/349463
+            return AuthenticationModes.Basic | AuthenticationModes.Pat;
+        }
+
+        // <remarks>Stores OAuth tokens as a side effect</remarks>
+        public override async Task<ICredential> GetCredentialAsync(InputArguments input)
+        {
+            string service = GetServiceName(input);
+            ICredential credential = Context.CredentialStore.Get(service, input.UserName);
+            if (credential?.Account == "oauth2" && await IsOAuthTokenExpired(input.GetRemoteUri(), credential.Password))
+            {
+                Context.Trace.WriteLine("Removing expired OAuth access token...");
+                Context.CredentialStore.Remove(service, credential.Account);
+                credential = null;
+            }
+
+            if (credential != null)
+            {
+                return credential;
+            }
+
+            string refreshService = GetRefreshTokenServiceName(input);
+            string refreshToken = Context.CredentialStore.Get(service, input.UserName)?.Password;
+            if (refreshToken != null)
+            {
+                Context.Trace.WriteLine("Refreshing OAuth token...");
+                try
+                {
+                    credential = await RefreshOAuthCredentialAsync(input, refreshToken);
+                }
+                catch (Exception e)
+                {
+                    Context.Terminal.WriteLine($"OAuth token refresh failed: {e.Message}");
+                }
+            }
+
+            credential ??= await GenerateCredentialAsync(input);
+
+            if (credential is OAuthCredential oAuthCredential)
+            {
+                Context.Trace.WriteLine("Pre-emptively storing OAuth access and refresh tokens...");
+                // freshly-generated OAuth credential
+                // store credential, since we know it to be valid (whereas Git will only store credential if git push succeeds)
+                Context.CredentialStore.AddOrUpdate(service, oAuthCredential.Account, oAuthCredential.AccessToken);
+                // store refresh token under a separate service
+                Context.CredentialStore.AddOrUpdate(GetRefreshTokenServiceName(input), oAuthCredential.Account, oAuthCredential.RefreshToken);
```

Comment:
This value has already been calculated.. no need to re-calculate it.

```suggestion
                Context.CredentialStore.AddOrUpdate(refreshService, oAuthCredential.Account, oAuthCredential.RefreshToken);
```